caracal/workers/selfcal_worker.py

```diff
@@ -360,6 +360,18 @@ def worker(pipeline, recipe, config):
             if not os.path.exists(os.path.join(pipeline.msdir, m)):
                 raise IOError(
                     "MS file {0:s}, to transfer gains to, does not exist. Please check that it is where it should be.".format(m))
+            # Check if a model subtraction has already been done
+            with table('{0:s}/{1:s}'.format(pipeline.msdir, m), readonly=False) as ms_table:
+                try:
+                    caracal.log.info(f"Checking if the model has been subtracted in {m}")
+                    nModelSub = ms_table.getcolkeyword('CORRECTED_DATA', 'modelSub')
```

Comment:
> That's just for logging purposes to show and document that the counter has been reset. It's not essential.

I'm not sure if I follow here. This is what I meant. These two lines follow each other, and it seems the first one is frivolous.
Unless you mean, calling the function `getcolkeyword` will log some results.
```
                    nModelSub = ms_table.getcolkeyword('CORRECTED_DATA', 'modelSub')
                    nModelSub = 0
```

> I'm not sure if it was this instance, but I think I needed to close the table, otherwise it would hang.

If you need to close the table, do it in a finally statement, since if the try doesn't succeed, that table might not close.
Finally block will always get executed, either with an exception generated or not

Reply:
Ahhh I understand what you mean now. This is because of how I was testing it. I guess at some stage, I had something more along the line of:

```
nModelSub = ms_table.getcolkeyword('CORRECTED_DATA', 'modelSub')
caracal.log.info(f"The model has been subtracted {nModelSub}, resetting...")
nModelSub = 0
```

So, it was first to see that a model subtraction **had** been done, so that I could see every step of the reset under these conditions. But I agree, the first one is superfluous, because if it's found the keyword it's going to reset it regardless. 

So, I can take `nModelSub = ms_table.getcolkeyword('CORRECTED_DATA', 'modelSub')` out, and put a finally in to close the table. I guess I should do this in all instances in the selflcal worker, not just at `trasfer_apply_gains`.

Reply:
> ... and put a finally in to close the table.

Yes, but no need.
Check this link about context managers.
https://book.pythontips.com/en/latest/context_managers.html

Reply:
So @Athanaseus 

Is this what you suggest?:

`            with table('{0:s}/{1:s}'.format(pipeline.msdir, m), readonly=False) as ms_table:
                try:
                    caracal.log.info(f"Checking if the model has been subtracted in {m}")
                    nModelSub = 0
                    ms_table.putcolkeyword('CORRECTED_DATA', 'modelSub', nModelSub)
                    nModelSub = ms_table.getcolkeyword('CORRECTED_DATA', 'modelSub')
                    caracal.log.info(f"Reseting the counter to {nModelSub}")
                except RuntimeError:
                    caracal.log.info(f"No subtraction found in {m}")
                finally:
                    ms_table.close()`

Reply:
Not quite, but this will work too.
My suggestion was either use `with table...`  or `finally`.
e.g.
```
            with table('{0:s}/{1:s}'.format(pipeline.msdir, m), readonly=False) as ms_table:
                try:
                    caracal.log.info(f"Checking if the model has been subtracted in {m}")
                    nModelSub = 0
                    ms_table.putcolkeyword('CORRECTED_DATA', 'modelSub', nModelSub)
                    caracal.log.info(f"Reseting the counter to {nModelSub}")
                except RuntimeError:
                    caracal.log.info(f"No subtraction found in {m}")
```

or 


```
            ms_table = table('{0:s}/{1:s}'.format(pipeline.msdir, m), readonly=False)
            try:
                caracal.log.info(f"Checking if the model has been subtracted in {m}")
                nModelSub = 0
                ms_table.putcolkeyword('CORRECTED_DATA', 'modelSub', nModelSub)
                caracal.log.info(f"Reseting the counter to {nModelSub}")
            except RuntimeError:
                caracal.log.info(f"No subtraction found in {m}")
            finally:
                ms_table.close()
```